tests/tensor/test_math.py

```diff
@@ -2202,12 +2205,79 @@ def test_broadcastable2(self):
         axes = [[2, 1], [0, 1]]
         z = tensordot(x, y, axes=axes)
         assert z.broadcastable == (True, False)
+        assert z.type.shape == (1, None)
         f = inplace_func([x, y], z)
         xv = random(1, 3, 4, rng=rng)
         yv = random(4, 3, 5, rng=rng)
         zv = f(xv, yv)
         assert np.allclose(np.tensordot(xv, yv, axes=axes), zv)
 
+    def test_type_shape(self):
+        x = ones(shape=(7, 3, 2))
+        y = ones(
+            shape=(
+                10,
+                2,
+            )
+        )
+        xv = x.eval()
+        yv = y.eval()
+        sy = tensor("sy", shape=(None, 2))
+        axes = [[-1], [-1]]
+        z = tensordot(x, y, axes=axes)
+        sz = tensordot(x, sy, axes=axes)
+
+        fg = FunctionGraph([x, y], [z])
+        assert not any(isinstance(n, Assert) for n in fg.toposort())
+        assert z.type.shape == (7, 3, 10)
+        assert z.broadcastable == (False, False, False)
+        assert np.allclose(np.tensordot(xv, yv, axes=axes), z.eval())
+
+        fg = FunctionGraph([x, sy], [sz])
+        assert not any(isinstance(n, Assert) for n in fg.toposort())
+        assert sz.type.shape == (7, 3, None)
+        assert z.broadcastable == (False, False, False)
+        assert np.allclose(np.tensordot(xv, yv, axes=axes), sz.eval({sy: yv}))
+
+        with pytest.raises(
+            ValueError,
+            match="Input arrays have inconsistent broadcastable pattern or type shape",
+        ):
+            tensordot(ones(shape=(7, 4)), ones(shape=(7, 4)), axes=1)
+
+    @pytest.mark.parametrize(
+        ["axes", "has_assert", "values", "expected_fail"],
+        [
+            ([[1], [2]], False, (np.ones((7, 3, 2)), np.ones((7, 2, 3))), False),
+            ([[1, 2], [2, 1]], True, (np.ones((7, 3, 2)), np.ones((7, 2, 3))), False),
+            ([[1, 2], [2, 1]], True, (np.ones((7, 3, 2)), np.ones((7, 5, 3))), True),
+        ],
+    )
+    def test_shape_assert(self, axes, has_assert, values, expected_fail):
+        x = tensor(shape=(7, 3, None))
+        y = tensor(shape=(None, None, 3))
+
+        xv, yv = values
+        xv = xv.astype(x.dtype)
+        yv = yv.astype(x.dtype)
+
+        # No assert should be present
+        z = tensordot(x, y, axes=axes)
+        fg = FunctionGraph([x, y], [z])
```

Comment:
Nitpick 

No need for fg, just check Ops in ancestors (maybe there's a utility already to return nodes from ancestors)

Reply:
Something like set(node.owner for node in ancestors(...) if node.owner) should do the trick

Reply:
But you can already filter /look for asserts